### ISSUE TITLE
Implementation of ensemble

### ISSUE THREAD

**Issue by xidulu** [Wang-ML-Lab/bayesian-peft#3]
Hi there

Nice work, is there any particular reason for sticking with the non-standard implementation of ensemble? i.e. averaging the logits then softmax rather than softmax followed by averaging?

https://github.com/Wang-ML-Lab/bayesian-peft/blob/main/modelwrappers/deepensemble.py#L183

And apperantly you do things the other way around when implementing the Bayesian model averaging for blob:

https://github.com/Wang-ML-Lab/bayesian-peft/blob/main/modelwrappers/blob.py#L575

Also the inference budget is different for ensemble and Blob, is there any reason behind?

For VI, you use 10 samples
https://github.com/Wang-ML-Lab/bayesian-peft/blob/main/scripts/blob/blob-llama-ood-all.sh#L10

For ensemble, you use 3 samples
https://github.com/Wang-ML-Lab/bayesian-peft/blob/main/scripts/deepensemble/deepensemble-llama-all.sh#L18

**Comment by haizhou-shi** [comment]
Hi @xidulu,

Thanks for your questions. 

**About Averaging Logits instead of Prob.** According to my vague memory when implementing the DeepEnsemble for LoRA, we tried both ways of ensemble, i.e., averaging the logits and averaging the final softmax outputs. The former slightly outperforms the latter and we decided to report the former implementation in the paper. But I will double check it since you bring this issue up. 

**About the budget for DeepEnsemble.** This is also a good question. DeepEnsemble with sample size = 10 requires 10x more memory and 10x more time or training (if trained sequentially). Compared to other baselines it would be too much and unfair to the other methods. Hence we pick an acceptable number of LoRAs during training (which is 3).

Have you tried (i) averaging the probs instead of logits for DeepEnsemble and (ii) setting the number of LoRAs to 10 with our codebase? If you want we can conduct some experiments of moderate sizes. 

Thank you for your comment!

**Comment by xidulu** [comment]
Thanks for the fast response. I raise the question mainly out of curiosity, it is a non-standard implementation, but I cannot think of any major issues with it, especially considering that you are averaging over normalized logits (i.e. the ones output by log_softmax) so you are taking the geometric mean rather than the arithmetic of the probability vector. 

> (i) averaging the probs instead of logits for DeepEnsemble

Not really, I've never performed averaging over logtis in my life, that's why your implementation makes me curious.

> (ii) setting the number of LoRAs to 10 with our codebase? 

Not really, but I believe it will certainly boost the ensemble's performance by a large margin.

**Comment by haizhou-shi** [comment]
Thank you for your comment!
